README.md
# WP Google Fonts Optimizer

[![PHP Version](https://img.shields.io/badge/php-5.4%2B-blue.svg?style=flat-square)](https://packagist.org/packages/zytzagoo/wp-google-fonts-optimizer)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/zytzagoo/wp-google-fonts-optimizer.svg?style=flat-square)](https://travis-ci.org/zytzagoo/wp-google-fonts-optimizer)
[![Scrutinizer Coverage](https://img.shields.io/scrutinizer/coverage/g/zytzagoo/wp-google-fonts-optimizer.svg?style=flat-square)](https://scrutinizer-ci.com/g/zytzagoo/wp-google-fonts-optimizer/?branch=master)
[![Quality Score](https://img.shields.io/scrutinizer/g/zytzagoo/wp-google-fonts-optimizer.svg?style=flat-square)](https://scrutinizer-ci.com/g/zytzagoo/wp-google-fonts-optimizer)

WP Google Fonts Optimizer is a super-easy way to ensure your WordPress theme is
not performing unnecessary extra requests for Google Web Fonts (in cases when
you use more than one font family on a page).

It automatically scans your enqueued stylesheets and combines them into a single request when needed.

Optionally, for poorly coded themes (and/or conflicts with other
plugins), it can scan and modify the generated markup.

## Download

See [Releases](https://github.com/zytzagoo/wp-google-fonts-optimizer/releases).

Or install it to your plugin directory via Composer:

```
composer create-project zytzagoo/wp-google-fonts-optimizer --no-dev
```

## Quickstart

Install and activate the plugin, it should do it's job automatically after that.

## Details / Troubleshooting

By default, the plugin enqueues a new stylesheet (with combined font families
etc.) and removes any found/enqueued stylesheets.

If your theme doesn't enqueue the Google Fonts properly (or if there is a
potential conflict with another plugin/theme on your site), you can modify the
way the plugin works and change it so that it parses the generated markup
(instead of it inspecting the enqueued URLs). Do so by adding a filter to
your `functions.php` (or even better, use a `mu-plugin`).

```php
add_filter( 'zwf_gfo_mode', function( $mode ) {
    return 'markup';
});
```

When in _markup mode_, it replaces existing `<link>` elements with a new one
and places it in the `<head>`. This mode also supports creating a
[Web Font Loader](https://github.com/typekit/webfontloader) `<script>` tag,
if that's what you'd prefer. Turn it on using a filter:

```php
add_filter( 'zwf_gfo_markup_type', function( $type ) {
    return 'script';
});
```

## [License (MIT)](LICENSE.md)


CHANGELOG.md
# Change Log

## [Unreleased]
### Added
- n/a

### Changed
- n/a

### Fixed
- n/a

## [0.2.0] - 2017-10-24
### Added
- AMP markup is supported and tested

### Changed
- Re-arranged some code/methods visibility (which breaks the public API in theory)

### Fixed
- Fixed some possible edge cases
- Fixed several tests

## [0.0.1] - 2017-10-12
### Added
- Initial release


tests/TestCase.php
<?php

namespace ZWF\GoogleFontsOptimizer\Tests;

use Brain\Monkey;

/**
 * Abstract base class for all test case implementations.
 *
 * @package ZWF\GoogleFontsOptimizer\Tests
 */
abstract class TestCase extends \PHPUnit_Framework_TestCase
{
    /**
     * Prepares the test environment before each test.
     *
     * @return void
     */
    protected function setUp()
    {
        parent::setUp();
        Monkey\setUp();
    }

    /**
     * Cleans up the test environment after each test.
     *
     * @return void
     */
    protected function tearDown()
    {
        Monkey\tearDown();
        parent::tearDown();
    }
}


tests/Unit/GoogleFontTest.php
<?php

namespace ZWF\GoogleFontsOptimizer\Tests\Unit;

use ZWF\GoogleFontsOptimizer\Tests\TestCase;
use ZWF\GoogleFont;

/**
 * Test cases for the GoogleFont class.
 *
 * @package ZWF\GoogleFontsOptimizer\Tests\Unit
 */
class GoogleFontTest extends TestCase
{
    public function testFromStringHelper()
    {
        $test = 'Ubuntu:400,700italic,400italic:cyrillic';
        $font = GoogleFont::fromString($test);

        $this->assertSame('Ubuntu', $font->getName());
        $this->assertSame(['400','400italic','700italic'], $font->getSizes());
        $this->assertSame('400,400italic,700italic', $font->getSizesString());
        $this->assertSame(['cyrillic'], $font->getSubsets());
        $this->assertSame('cyrillic', $font->getSubsetsString());
        $this->assertSame('Ubuntu:400,400italic,700italic:cyrillic', (string) $font);

        $test = 'Roboto+Condensed:400';
        $font = GoogleFont::fromString($test);

        $this->assertSame('Roboto+Condensed', $font->getName());
        $this->assertSame(['400'], $font->getSizes());
        $this->assertSame('400', $font->getSizesString());
        $this->assertSame([], $font->getSubsets());
        $this->assertSame('', $font->getSubsetsString());
        $this->assertSame('Roboto+Condensed:400', (string) $font);

        $test = 'Roboto';
        $font = GoogleFont::fromString($test);

        $this->assertSame('Roboto', $font->getName());
        $this->assertSame([], $font->getSizes());
        $this->assertSame('', $font->getSizesString());
        $this->assertSame([], $font->getSubsets());
        $this->assertSame('', $font->getSubsetsString());
        $this->assertSame('Roboto', (string) $font);
    }

    /**
     * @expectedException \InvalidArgumentException
     */
    public function testException()
    {
        $test = '';
        $font = GoogleFont::fromString($test);
    }
}


tests/Unit/GoogleFontsOptimizerTest.php
<?php

namespace ZWF\GoogleFontsOptimizer\Tests\Unit;

use Brain\Monkey;
use ZWF\GoogleFontsOptimizer as Testee;
use ZWF\GoogleFontsOptimizerUtils as Utils;
use ZWF\GoogleFontsOptimizer\Tests\TestCase;

/**
 * Test case for the GoogleFontsOptimizer class.
 *
 * @package ZWF\GoogleFontsOptimizer\Tests\Unit
 */
class GoogleFontsOptimizerTest extends TestCase
{
    public function testStartBufferingShouldBuffer()
    {
        $instance = new Testee();

        // Starts
        $this->assertTrue($instance->startBuffering());
        $contents = ob_get_clean(); // otherwise the test ends up being marked as risky

        // Doesn't start
        define('WP_CLI', true);
        $this->assertFalse($instance->startBuffering());
    }

    public function testEndBuffering()
    {
        $instance = new Testee();

        // No changes to markup
        $markup   = 'bla';
        $expected = 'bla';
        $actual   = $instance->endBuffering($markup);
        $this->assertSame($expected, $actual);

        // No changes when only one link in markup
        $markup   = <<<HTML
<html>
    <head>
    <link href="//fonts.googleapis.com/css?family=Open+Sans:400italic,700italic,400,700" rel="stylesheet" type="text/css">
</head>
<body>
</body>
</html>
HTML;
        $expected = $markup;
        $actual   = $instance->endBuffering($markup);
        $this->assertSame($expected, $actual);

        // Actually replaces stuff when there is stuff to replace...
        $markup   = <<<HTML
<html>
    <head>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:400italic,700italic,400,700' rel='stylesheet' type='text/css'>
    <link href='http://fonts.googleapis.com/css?family=Ubuntu:400,700,400italic,700italic&subset=latin,latin-ext,cyrillic' rel='stylesheet' type='text/css'>
    <link href='https://fonts.googleapis.com/css?family=Raleway:400,700&amp;subset=latin,latin-ext' rel='stylesheet' type='text/css'>
    <link href='http://fonts.googleapis.com/css?family=Roboto+Condensed&text=Woot' rel='stylesheet' type='text/css'>
</head>
<body>
</body>
</html>
HTML;
        $expected = <<<HTML
<html>
    <head><link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Open+Sans%3A400%2C400italic%2C700%2C700italic%7CRaleway%3A400%2C700%7CUbuntu%3A400%2C400italic%2C700%2C700italic&amp;subset=latin%2Clatin-ext%2Ccyrillic"><link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto+Condensed&amp;text=Woot">
    </head>
<body>
</body>
</html>
HTML;

        $actual = $instance->endBuffering($markup);
        $this->assertSame($expected, $actual);
    }

    /**
     * Test processStylesHandles() works as expected
     */
    public function testProcessStylesHandlesDefaults()
    {
        $test_handles = [
            'my-fonts' => 'http://fonts.googleapis.com/css?family=Raleway:300,500,700&subset=latin,latin-ext',
            'my-font-text' => 'https://fonts.googleapis.com/css?family=Inconsolata&text=Hello',
            'my-font-text-spaces' => 'https://fonts.googleapis.com/css?family=Open+Sans&text=Hello',
            'my-font-text-spaces-2' => 'https://fonts.googleapis.com/css?family=Roboto+Condensed&text=Woot',
            'multi' => 'http://fonts.googleapis.com/css?family=Open+Sans:300|PT+Sans:300,400,700&subset=latin,cyrillic',
            'wfl-syntax' => 'http://fonts.googleapis.com/css?family=Ubuntu:300,700:latin,latin-ext|Roboto:400,700:cyrillic',
            'wfl-syntax-2' => 'http://fonts.googleapis.com/css?family=Ubuntu:600:latin|Roboto:500:cyrillic',
            'some-duplicates-in-the-mix' => 'http://fonts.googleapis.com/css?family=Ubuntu:300|Roboto:400&subset=latin,cyrillic'
        ];

        $expected_handles = [
            'zwf-gfo-combined' => 'https://fonts.googleapis.com/css?family=Open+Sans%3A300%7CPT+Sans%3A300%2C400%2C700%7CRaleway%3A300%2C500%2C700%7CRoboto%3A400%2C500%2C700%7CUbuntu%3A300%2C600%2C700&subset=latin%2Clatin-ext%2Ccyrillic',
            'zwf-gfo-combined-txt-1' => 'https://fonts.googleapis.com/css?family=Inconsolata&text=Hello',
            'zwf-gfo-combined-txt-2' => 'https://fonts.googleapis.com/css?family=Open+Sans&text=Hello',
            'zwf-gfo-combined-txt-3' => 'https://fonts.googleapis.com/css?family=Roboto+Condensed&text=Woot'
        ];

        // Mock methods relying on WP core functions
        // @codingStandardsIgnoreLine
        $stub = $this->getMockBuilder(Testee::class)
            ->setMethods(['findCandidateHandles'])
            ->getMock();

        // Needs to have its behavior changed since there's no WP core here
        $stub->method('findCandidateHandles')->willReturn($test_handles);
        // findCandidateHandles called with what we expect
        $stub->expects($this->once())
            ->method('findCandidateHandles')
            ->with($this->equalTo(array_keys($test_handles)));

        // We need to set candidates manually since findCandidateHandles is mocked
        $stub->setCandidates(array_values($test_handles));

        // Finally, call processStylesHandles with data that would match the
        // way it would be called when it's hooked in WP's `print_styles_array`
        $results = $stub->processStylesHandles(array_keys($test_handles));

        // Making sure we return handle names only, as would be done in WP-land
        $this->assertSame(array_keys($expected_handles), array_values($results));

        // Ensure the newly enqueued urls are what we expect them to be
        $list = $stub->getEnqueued();
        $this->assertSame(count($expected_handles), count($list));
        foreach ($expected_handles as $handle => $url) {
            $this->assertSame($url, $stub->getEnqueued($handle));
        }
    }

    /**
     * Test processStylesHandles() doesn't do anything when it shouldn't have to
     */
    public function testProcessStylesHandlesWithoutGoogleFonts()
    {
        $test_handles     = [
            'my-fonts' => 'http://www.example.org/fooo.css'
        ];
        $expected_handles = [
            'my-fonts' => 'http://www.example.org/fooo.css'
        ];

        // Mock methods relying on WP core functions
        // @codingStandardsIgnoreLine
        $stub = $this->getMockBuilder(Testee::class)
            ->setMethods(['findCandidateHandles', 'dequeueStyleHandles', 'enqueueStyle'])
            ->getMock();

        // Needs to have its behavior changed since there's no WP core here
        $stub->method('findCandidateHandles')->willReturn([]);

        // Ensure findCandidateHandles is called with what we expect
        $stub->expects($this->once())
            ->method('findCandidateHandles')
            ->with($this->equalTo(array_keys($test_handles)));
        // These methods shouldn't be called
        $stub->expects($this->never())
            ->method('dequeueStyleHandles');
        $stub->expects($this->never())
            ->method('enqueueStyle');

        // Finally, call processStylesHandles with data that would match the
        // way it would be called when it's hooked via WP's `print_styles_array`
        $results = $stub->processStylesHandles(array_keys($test_handles));
        $this->assertSame(array_keys($expected_handles), array_values($results));
    }

    /**
     * Test processMarkup using AMP
     */
    public function testProcessMarkupWithAmp()
    {
        $input = <<<MARKUP
<!doctype html>
<html amp lang="en">
    <head>
    <meta charset="utf-8">
    <script async src="https://cdn.ampproject.org/v0.js"></script>
    <title>Hello, AMPs</title>
    <link href="https://fonts.googleapis.com/css?family=Tangerine" rel="stylesheet">
    <link href="http://fonts.googleapis.com/css?family=Bitstream+Vera+Serif" rel="stylesheet">
    <link rel="canonical" href="http://example.ampproject.org/article-metadata.html">
    <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
    </head>
    <body>
    <h1>Welcome to the mobile web</h1>
    </body>
</html>
MARKUP;

        $expected = <<<MARKUP
<!doctype html>
<html amp lang="en">
    <head><link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Bitstream+Vera+Serif%7CTangerine">
    <meta charset="utf-8">
    <script async src="https://cdn.ampproject.org/v0.js"></script>
    <title>Hello, AMPs</title>
    <link rel="canonical" href="http://example.ampproject.org/article-metadata.html">
    <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
    </head>
    <body>
    <h1>Welcome to the mobile web</h1>
    </body>
</html>
MARKUP;

        $instance = new Testee();
        $actual   = $instance->processMarkup($input);

        $this->assertSame($expected, $actual);
    }
}


index.php
<?php
// @codingStandardsIgnoreFile
/**
 * Plugin Name:  WP Google Fonts Optimizer
 * Plugin URI:   https://wordpress.org/plugins/wp-google-fonts-optimizer/
 * Description:  Automatically detect and combine multiple Google Web Fonts requests into a single one.
 * Author:       zytzagoo
 * Author URI:   https://zytzagoo.net
 * Version:      0.2.0
 * Requires PHP: 5.4
 *
 * @package ZWF\GoogleFontsOptimizer
 */

namespace ZWF;

defined( 'ABSPATH' ) || exit;

/**
 * Bootstraps the plugin.
 *
 * @wp-hook plugins_loaded
 *
 * @return void
 */
function bootstrap() {
    if ( file_exists( __DIR__ . '/vendor/autoload.php' ) ) {
        // Composer-generated autoload file.
        include_once __DIR__ . '/vendor/autoload.php';
    }

    $combiner = new GoogleFontsOptimizer();
    add_action( 'wp', [ $combiner, 'run' ] );
}

add_action( 'plugins_loaded', __NAMESPACE__ . '\\bootstrap', 0 );


src/GoogleFont.php
<?php

namespace ZWF;

class GoogleFont
{

    /**
     * @var string
     */
    protected $name = '';

    /**
     * @var array
     */
    protected $sizes = [];

    /**
     * @var array
     */
    protected $subsets = [];

    public function __construct($name, array $sizes = [], array $subsets = [])
    {
        $this->setName($name);
        $this->setSizes($sizes);
        $this->setSubsets($subsets);
    }

    /**
     * @param string $name
     * @throws \InvalidArgumentException
     */
    public function setName($name)
    {
        if (empty($name)) {
            throw new \InvalidArgumentException('Font `$name` is required!');
        }

        $this->name = $name;
    }

    /**
     * @return string
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * @param string|array $sizes
     */
    public function setSizes($sizes)
    {
        $this->stringOrArraySetter('sizes', $sizes);

        $this->sizes = array_unique($this->sizes);
        sort($this->sizes, SORT_REGULAR);
    }

    /**
     * @return array
     */
    public function getSizes()
    {
        return $this->sizes;
    }

    /**
     * @return string
     */
    public function getSizesString()
    {
        return implode(',', $this->getSizes());
    }

    /**
     * @param string|array $subsets
     */
    public function setSubsets($subsets)
    {
        $this->stringOrArraySetter('subsets', $subsets);

        $this->subsets = array_unique($this->subsets, SORT_STRING);
    }

    /**
     * @return array
     */
    public function getSubsets()
    {
        return $this->subsets;
    }

    /**
     * @return string
     */
    public function getSubsetsString()
    {
        return implode(',', $this->subsets);
    }

    /**
     * @return string
     */
    public function __toString()
    {
        $parts   = [
            $this->getName()
        ];
        $sizes   = $this->getSizesString();
        $subsets = $this->getSubsetsString();

        if (! empty($sizes)) {
            $parts[] = $sizes;
        }
        if (! empty($subsets)) {
            $parts[] = $subsets;
        }

        return implode(':', $parts);
    }

    /**
     * Setter for given `$property` which makes sure that if a string `$value`
     * is given (in which multiple values can be separated by a comma) it ends
     * up being converted into an array.
     *
     * @param string $property
     * @param string|array $value
     *
     * @return void
     */
    private function stringOrArraySetter($property, $value)
    {
        if (is_array($value)) {
            $this->$property = $value;
        }
        if (is_string($value)) {
            $this->$property = array_map('trim', explode(',', $value));
        }
    }

    /**
     * Creates a new instance from a given $fontstring.
     *
     * @param string $fontstring
     *
     * @return GoogleFont
     */
    public static function fromString($fontstring)
    {
        $parts = explode(':', $fontstring);
        $font  = new self($parts[0]);

        if (isset($parts[1])) {
            $font->setSizes($parts[1]);
        }

        if (isset($parts[2])) {
            $font->setSubsets($parts[2]);
        }

        return $font;
    }
}


src/GoogleFontsCollection.php
<?php

namespace ZWF;

use ZWF\GoogleFontsOptimizerUtils as Utils;

class GoogleFontsCollection
{

    protected $links = [];

    protected $texts = [];

    /**
     * @var array|ZWF\GoogleFont[]
     */
    protected $fonts = [];

    protected $subsetsMap = [];

    protected $namedSizes = [];

    public function __construct(array $urls = [])
    {
        if (! empty($urls)) {
            foreach ($urls as $url) {
                $this->add($url);
            }
        }
    }

    public function add($url)
    {
        $this->addLink($url);

        $params = [];
        parse_str(parse_url($url, PHP_URL_QUERY), $params);

        $bucket = isset($params['text']) ? 'text' : 'complete';

        $fontstrings = $this->buildFontStringsFromQueryParams($params);
        foreach ($fontstrings as $font) {
            $this->addFontFromString($font, $bucket);
            if ('text' === $bucket) {
                $font_family = explode(':', $font);
                $this->addTextFont($font_family[0], Utils::httpsify($url));
            }
        }
    }

    protected function addLink($url)
    {
        $this->links[] = $url;
    }

    protected function addFontFromString($fontstring, $bucket = 'complete')
    {
        $font    = GoogleFont::fromString($fontstring);
        $name    = $font->getName();
        $subsets = $font->getSubsetsString();

        // Add to bucket list
        $this->fonts[$bucket][] = $font;

        // Keeping a separate map of names => subsets for webfontloader purposes
        if (! array_key_exists($name, $this->subsetsMap)) {
            // Nothing found yet, create a new key
            $this->subsetsMap[$name] = $subsets;
        } else {
            // Something is in there already, merge
            $this->subsetsMap[$name] = array_merge($this->subsetsMap[$name], (array) $subsets);
        }

        // Deduplicate values and don't sort the subsets map
        $this->subsetsMap = Utils::dedupValues($this->subsetsMap, false); // no sort

        // Maintain the list/hash of name => sizes when adding a new one too
        $this->namedSizes = $this->buildNamedsizesMap();
    }

    protected function addTextFont($name, $url)
    {
        $this->texts['name'][] = $name;
        $this->texts['url'][]  = $url;
    }

    /**
     * Returns an array of data needed to generated webfontloader script markup.
     *
     * @return array
     */
    public function getScriptData()
    {
        return [
            $this->getNamedSizesMap(),
            $this->getSubsetsMap()
        ];
    }

    public function getNamedSizesMap()
    {
        return $this->namedSizes;
    }

    /**
     * @return array
     */
    protected function getSubsetsMap()
    {
        return $this->subsetsMap;
    }

    public function getOriginalLinks()
    {
        return $this->links;
    }

    public function hasText()
    {
        return (! empty($this->texts));
    }

    /**
     * @return array
     */
    public function getTextUrls()
    {
        $urls = [];

        if ($this->hasText()) {
            $urls = $this->texts['url'];
        }

        return $urls;
    }

    /**
     * @return array
     */
    public function getTextNames()
    {
        return $this->texts['name'];
    }

    /**
     * @return ZWF\GoogleFont[]
     */
    public function getFonts()
    {
        return $this->fonts['complete'];
    }

    /**
     * @return string
     */
    public function getCombinedUrl()
    {
        return Utils::buildGoogleFontsUrl(
            $this->getNamedSizesMap(),
            $this->getSubsets()
        );
    }

    /**
     * @return array
     */
    public function getSubsets()
    {
        $subsets = [];

        $fonts = $this->getFonts();
        foreach ($fonts as $font) {
            $subsets[] = $font->getSubsets();
        }

        $subsets = Utils::arrayFlattenIterative($subsets);

        return array_unique($subsets);
    }

    /**
     * @return array
     */
    protected function buildNamedsizesMap()
    {
        $fonts = [];

        foreach ($this->getFonts() as $font) {
            $name = $font->getName();
            if (! array_key_exists($name, $fonts)) {
                $fonts[$name] = $font->getSizes();
            } else {
                $fonts[$name] = array_merge($fonts[$name], $font->getSizes());
            }
        }

        // Sanitize and de-dup
        $fonts = Utils::dedupValues($fonts, SORT_REGULAR); // sorts values (sizes)

        // Sorts by font names alphabetically
        ksort($fonts);

        return $fonts;
    }

    /**
     * Looks for and parses the `family` key in `$params` array to get back
     * the list of fonts/sizes/subsets requested via a GoogleFonts url.
     *
     * @param array $params
     *
     * @return array
     */
    protected function buildFontStringsFromQueryParams(array $params)
    {
        $fonts = [];

        foreach (explode('|', $params['family']) as $family) {
            $font = $this->parseFontStringName($family, $params);
            if (! empty($font)) {
                $fonts[] = $font;
            }
        }

        return $fonts;
    }

    /**
     * @param string $family
     * @param array $params
     *
     * @return string
     */
    protected function parseFontStringName($family, array $params)
    {
        $subset = false;
        $family = explode(':', $family);

        if (isset($params['subset'])) {
            // Use the found subset query parameter
            $subset = $params['subset'];
        } elseif (isset($family[2])) {
            // Use the subset in the family string if present
            $subset = $family[2];
        }

        // $family can have a lot of thing specified with separators etc.
        $parts = $this->validateFontStringParts($family, $subset);
        $font  = implode(':', $parts);

        return $font;
    }

    /**
     * Validate and return desired font name/size/subset parts.
     *
     * @param array $family
     * @param bool $subset
     *
     * @return array
     */
    protected function validateFontStringParts(array $family, $subset = false)
    {
        $parts = [];

        // First part is the font name, which should always be present
        $parts[] = $family[0];

        // Check if sizes are specified
        if (isset($family[1]) && strlen($family[1]) > 0) {
            $parts[] = $family[1];
        }

        // Add the subset if specified
        if ($subset) {
            $parts[] = $subset;
        }

        return $parts;
    }
}


src/GoogleFontsOptimizerUtils.php
<?php

namespace ZWF;

use ZWF\GoogleFontsCollection as Collection;

class GoogleFontsOptimizerUtils
{
    /**
     * Replaces any occurrences of un-encoded ampersands in the given string
     * with the value given in the `$amp` parameter (`&amp;` by default).
     *
     * @param string $url
     * @param string $amp
     *
     * @return string
     */
    public static function encodeUnencodedAmpersands($url, $amp = '&amp;')
    {
        $amp = trim($amp);

        if (empty($amp)) {
            $amp = '&amp;';
        }

        return preg_replace('/&(?!#?\w+;)/', $amp, $url);
    }

    /**
     * Replaces protocol-relative or non-https URLs into https URLs.
     *
     * @param string $link
     *
     * @return string
     */
    public static function httpsify($link)
    {
        $is_protocol_relative = ('/' === $link{0} && '/' === $link{1});

        if ($is_protocol_relative) {
            $link = 'https:' . $link;
        } else {
            $link = preg_replace('/^(https?):\/\//mi', '', $link);
            $link = 'https://' . $link;
        }

        return $link;
    }

    /**
     * Returns true if a given `$url` is a Google Fonts URL.
     *
     * @param string $url
     *
     * @return bool
     */
    public static function isGoogleWebFontUrl($url)
    {
        return (substr_count($url, 'fonts.googleapis.com/css') > 0);
    }

    /**
     * Returns true if given `$string` contains the HTML5 doctype.
     *
     * @param string $string
     *
     * @return bool
     */
    public static function hasHtml5Doctype($string)
    {
        return (preg_match('/^<!DOCTYPE.+html>/i', $string) > 0);
    }

    /**
     * Returns true when given `$string` contains an XSL stylesheet element.
     *
     * @param string $string
     *
     * @return bool
     */
    public static function hasXslStylesheet($string)
    {
        return (false !== stripos($string, '<xsl:stylesheet'));
    }

    /**
     * Returns true when given `$string` contains the beginnings of an `<html>` tag.
     *
     * @param string $string
     * @return bool
     */
    public static function hasHtmlTag($string)
    {
        return (false !== stripos($string, '<html'));
    }

    /**
     * Given a `['key' => 'value1,value2,value3']` or a
     * `['key' => ['value1', 'value2', 'value3']]` map/array it returns a new
     * array with the same keys, but the values are now always an array of
     * values (and any potential duplicate values are removed).
     * If the `$sort` parameter is given, the list of values is sorted using
     * `sort()` and the `$sort` parameter is treated as a sort flag.
     *
     * @param array $data
     * @param bool|int $sort If false, no sorting, otherwise an integer representing
     *                       sort flags. See http://php.net/sort
     *
     * @return array
     */
    public static function dedupValues(array $data, $sort = false)
    {
        foreach ($data as $key => $values) {
            if (is_array($values)) {
                $parts = $values;
            } else {
                $parts = explode(',', $values);
            }

            $parts = array_unique($parts);

            // Perform sort if specified
            if (false !== $sort) {
                sort($parts, (int) $sort);
            }

            // Store back
            $data[$key] = $parts;
        }

        return $data;
    }

    /**
     * Given a `GoogleFontsCollection` it builds needed `<link rel="stylesheet">` markup.
     *
     * @param Collection $fonts
     *
     * @return string
     */
    public static function buildFontsMarkupLinks(Collection $fonts)
    {
        $font_url = $fonts->getCombinedUrl();
        $href     = self::encodeUnencodedAmpersands($font_url);
        $markup   = '<link rel="stylesheet" type="text/css" href="' . $href . '">';

        foreach ($fonts->getTextUrls() as $url) {
            $markup .= '<link rel="stylesheet" type="text/css"';
            $markup .= ' href="' . self::encodeUnencodedAmpersands($url) . '">';
        }

        return $markup;
    }

    /**
     * Given a `GoogleFontsCollection` it builds the WebFont loader script markup.
     *
     * @param Collection $collection
     *
     * @return string
     */
    public static function buildFontsMarkupScript(Collection $collection)
    {
        $families_array = [];

        list($names, $mapping) = $collection->getScriptData();

        foreach ($names as $name => $sizes) {
            $family = $name . ':' . implode(',', $sizes);
            if (isset($mapping[$name])) {
                $family .= ':' . implode(',', $mapping[$name]);
            }
            $families_array[] = $family;
        }
        $families = "'" . implode("', '", $families_array) . "'";

        // Load 'text' requests with the "custom" module
        $custom = '';
        if ($collection->hasText()) {
            $custom  = ",\n    custom: {\n";
            $custom .= "        families: [ '" . implode("', '", $collection->getTextNames()) . "' ],\n";
            $custom .= "        urls: [ '" . implode("', '", $collection->getTextUrls()) . "' ]\n";
            $custom .= '    }';
        }

        $markup = <<<MARKUP
<script type="text/javascript">
WebFontConfig = {
    google: { families: [ {$families} ] }{$custom}
};
(function() {
    var wf = document.createElement('script');
    wf.src = ('https:' == document.location.protocol ? 'https' : 'http') +
        '://ajax.googleapis.com/ajax/libs/webfont/1/webfont.js';
    wf.type = 'text/javascript';
    wf.async = 'true';
    var s = document.getElementsByTagName('script')[0];
    s.parentNode.insertBefore(wf, s);
})();
</script>
MARKUP;

        return $markup;
    }

    /**
     * Builds a combined Google Font URL for multiple font families/subsets.
     *
     * Usage examples:
     * ```
     * ZWF\GoogleFontsOptimizerUtils::buildGoogleFontsUrl(
     *     [
     *         'Open Sans' => [ '400', '400italic', '700', '700italic' ],
     *         'Ubuntu'    => [ '400', '400italic', '700', '700italic' ],
     *     ),
     *     [ 'latin', 'latin-ext' ]
     * );
     * ```
     *
     * or
     *
     * ```
     * ZWF\GoogleFontsOptimizerUtils::buildGoogleFontsUrl(
     *     [
     *         'Open Sans' => '400,400italic,700,700italic',
     *         'Ubuntu'    => '400,400italic,700,700italic',
     *     ],
     *     'latin,latin-ext'
     * );
     * ```
     *
     * @param array $fonts
     * @param array|string $subsets
     *
     * @return null|string
     */
    public static function buildGoogleFontsUrl(array $fonts, $subsets = [])
    {
        $base_url  = 'https://fonts.googleapis.com/css';
        $font_args = [];
        $family    = [];

        $fonts = self::dedupValues($fonts, false); // no sort
        foreach ($fonts as $font_name => $font_weight) {
            // Trimming end colon handles edge case of being given an empty $font_weight
            $family[] = trim(trim($font_name) . ':' . implode(',', $font_weight), ':');
        }

        $font_args['family'] = implode('|', $family);

        if (! empty($subsets)) {
            if (is_array($subsets)) {
                $subsets = array_unique($subsets);
                $subsets = implode(',', $subsets);
            }
            $font_args['subset'] = trim($subsets);
        }

        $url = $base_url . '?' . http_build_query($font_args);

        return $url;
    }

    /**
     * Flatten an array without recursion.
     *
     * @param array $arr
     *
     * @return array
     */
    public static function arrayFlattenIterative(array $arr)
    {
        $flat  = [];
        $stack = array_values($arr);

        while (! empty($stack)) {
            $value = array_shift($stack);
            if (is_array($value)) {
                $stack = array_merge(array_values($value), $stack);
            } else {
                $flat[] = $value;
            }
        }

        return $flat;
    }
}
